### ISSUE TITLE
Add Valgrind to Travis CI

### ISSUE THREAD

**Issue by desyncr** [zpm-project/zpm-zsh#22]
http://valgrind.org/docs/manual/quick-start.html

**Comment by fennecdjay** [comment]
Well, not specifically for travis, but [this](https://gist.github.com/fennecdjay/57970d9e04ba19043a118ec10cdd57a3) could make us able to check memory defaults in a breeze.

**Comment by fennecdjay** [comment]
Is it ok to make `zpm_memory.sh` executable?
and to have something like
```sh
export PATH=$PWD/util:$PATH
```
in `tests/setup.sh` or `Makefile`'s test target?

**Comment by desyncr** [comment]
I'd put it on a make target:

    make valgrind

I guess we can set up a 'zpm' function/alias rather than using sed.

**Comment by fennecdjay** [comment]
it looks like alias will work 🍾 
should we add a *header* in every test file?
```c
[ -z "$ZPM_MEMORY" ] && alias zpm="$ZPM_MEMORY"
```

then
```makefile
valgrind:
    export ZPM_MEMORY="$PWD/tests/zpm_memory.sh
    make test
    unset ZPM_MEMORY
```

changing test files would be easy:
```sh
for test_file in *.t                               :(
do echo -e "[ -z \"\$ZPM_MEMORY\" ] && alias zpm=\"\$ZPM_MEMORY\"\n$(cat $test_file)" > "$test_file"
done
```

what do you think?

**Comment by desyncr** [comment]
I was thinking on adding it to the `tests/setup.sh`. I'd also use Travis CI's [build matrix feature](https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix).

[Here's](https://gist.github.com/desyncr/5eccf82398a4ef1aaa0e2021247e80f9) an (untested) patch. Let me know what you think!

**Comment by fennecdjay** [comment]
> Here's an (untested) patch. Let me know what you think!
Nifty!

**Comment by desyncr** [comment]
Want to take over this task? I have done the Travis CI part and you seem more competent with `valgrind` and `c` in general.

**Comment by fennecdjay** [comment]
That fine 😺 
I'll work on a PR as soon as I finish the huge Gwion refactor task I'm in right now.

**Comment by fennecdjay** [comment]
I've got something.
see #27.

**Comment by desyncr** [comment]
Gonna take a look at it tonight! Thanks!

**Comment by fennecdjay** [comment]
Can be closed as #27 has been merged. :smile:

**Comment by desyncr** [comment]
Closing this!
